### PR TITLE
New buffer api

### PR DESCRIPTION
**Pull request by abergeron**
This includes #174, which should be merged first.

This rework is more significant and does cause changes to have to be made downstream, but after this, the API should be much more stable.

### CODE REVIEW

File: pygpu/gpuarray.pyx [hunk 1]
```diff
@@ -955,40 +935,33 @@ cdef class GpuContext:
...
             else:
-                raise get_exc(err), self.ops.ctx_error(NULL) + ": " + str(devno)
+                raise get_exc(err), gpucontext_error(NULL, err) + ": " + str(devno)
 
     def __enter__(self):
         if cuda_enter == NULL:
             raise RuntimeError("cuda_enter not available")
         if cuda_exit == NULL:
             raise RuntimeError("cuda_exit not available")
-        if self.ops != gpuarray_get_ops("cuda"):
```
Comment (nouiz) [inline pygpu/gpuarray.pyx]: You still need a check like this one here.

File: pygpu/gpuarray.pyx [hunk 2]
```diff
@@ -1517,8 +1488,11 @@ cdef class GpuArray:
         """
         return pygpu_copy(self, to_ga_order(order))
 
-    def transfer(self, GpuContext new_ctx, share=False):
-        return pygpu_transfer(self, new_ctx, share)
+    def transfer(self, GpuContext new_ctx):
+        cdef GpuArray r
+        r = pygpu_empty(self.ga.nd, self.ga.dimensions, self.ga.typecode,
+                        GA_C_ORDER, new_ctx, None)
```
Comment (nouiz) [inline pygpu/gpuarray.pyx]: If the source if f contiguous?

File: pygpu/gpuarray.pyx (hunk 1 above)
Comment (abergeron) [inline pygpu/gpuarray.pyx]: ok

File: pygpu/gpuarray.pyx (hunk 2 above)
Comment (abergeron) [inline pygpu/gpuarray.pyx]: I've added a check and a case for that.

File: pygpu/gpuarray.pxd [hunk 3]
```diff
@@ -142,6 +142,7 @@ cdef extern from "gpuarray/array.h":
 
     bint GpuArray_CHKFLAGS(_GpuArray *a, int fl)
     bint GpuArray_ISONESEGMENT(_GpuArray *a)
+    bint GpuArray_IS_C_CONTIGUOUS(_GpuArray *a)
```
Comment (nouiz) [inline pygpu/gpuarray.pxd]: Can you implement IS_F_CONTIGUOUS at the same time?

Comment (abergeron) [inline pygpu/gpuarray.pxd]: It is already implemented, but we don't need it for the python module.

**Comment by nouiz** [comment]
Except for the last commit, all the commis are in gh-178. I would make a new PR with just the last commit and close this PR. Then we can merge gh-178 and https://github.com/Theano/Theano/pull/4542. Then after we can merge the blas change with another bump in the version number with a new corresponding Theano PR.

**Comment by abergeron** [comment]
I would rather rebase #178 with everything if you want to do this way,

File: pygpu/gpuarray.pxd (hunk 3 above)
Comment (nouiz) [inline pygpu/gpuarray.pxd]: Why we don't need it in the python module? I understand that currently it won't be used, but if someone make a new Theano op and need this, I don't want him to need to make change to libgpuarray.

**Comment by nouiz** [comment]
ok. the only commit here also break the interface, so doing all this in one PR would only break the interface once.

File: pygpu/gpuarray.pxd (hunk 3 above)
Comment (abergeron) [inline pygpu/gpuarray.pxd]: In Theano we can use the C function from libgpuarray directly.

**Comment by abergeron** [comment]
Integrated into #178.

**Pull request closed without merging**